[TT-16831] docs: add OpenTelemetry metrics pages and restructure metrics section (#1808)
TykTechnologies/tyk-docs: api-management/traces/jaeger.mdx

---
title: "Send Tyk Traces to Jaeger"
description: "Step-by-step guide to connect Tyk Gateway distributed traces to Jaeger using Docker or Kubernetes."
keywords: "Distributed Tracing, Jaeger, OpenTelemetry, OTLP, Tyk Gateway, Kubernetes, Docker, Observability"
sidebarTitle: "Jaeger"
---

## Using Docker

> This quick start guide offers a detailed, step-by-step walkthrough for configuring Tyk API Gateway (OSS, self-managed or hybrid gateway connected to Tyk Cloud) with OpenTelemetry and [Jaeger](https://www.jaegertracing.io/) to significantly improve API observability. We will cover the installation of essential components, their configuration, and the process of ensuring seamless integration.

Reviewer: ```suggestion
This guide explains how to set up [Jaeger](https://www.jaegertracing.io/)  to ingest OpenTelemetry traces via the OpenTelemetry Collector (OTel Collector) using Docker. We will cover the installation of essential components, their configuration, and the process of ensuring seamless integration.
```

        - redis

    redis:
      image: redis:4.0-alpine
      ports:
        - 6379:6379
      command: redis-server --appendonly yes
    ```

    - Navigate to the directory containing the docker-compose.yml file in your terminal.
    - Execute the following command to start the services:

    ```bash
    docker compose up
    ```

4. **Explore OpenTelemetry Traces in Jaeger**

    - Start by sending a few requests to the API endpoint configured in Step 2:
    ```bash
    curl http://localhost:8080/hello-world/ -i
    ```

    - Access Jaeger at [http://localhost:16686](http://localhost:16686).
    - In Jaeger's interface:
    - Select the service named tyk-gateway.
    - Click the *Find Traces* button.

    You should observe traces generated by Tyk Gateway, showcasing the distributed tracing information.

    <img src="/img/distributed-tracing/opentelemetry/api-gateway-trace-tyk-jaeger.png" alt="Tyk API Gateway distributed trace in Jaeger" />

    Select a trace to visualize its corresponding internal spans:

    <img src="/img/distributed-tracing/opentelemetry/api-gateway-trace-tyk-jaeger-spans.png" alt="Tyk API Gateway spans in Jaeger" />

## Using Kubernetes

> This quick start guide offers a detailed, step-by-step walkthrough for configuring Tyk Gateway OSS with OpenTelemetry and [Jaeger](https://www.jaegertracing.io/) on Kubernetes to significantly improve API observability. We will cover the installation of essential components, their configuration, and the process of ensuring seamless integration.

Reviewer: ```suggestion
This guide explains how to set up [Jaeger](https://www.jaegertracing.io/)  to ingest OpenTelemetry traces via the OpenTelemetry Collector (OTel Collector) using Kubernetes. We will cover the installation of essential components, their configuration, and the process of ensuring seamless integration.
```

    To install or upgrade [Tyk Gateway OSS using Helm](https://github.com/TykTechnologies/tyk-charts/tree/main/tyk-oss), execute the following commands:

    ```bash
    NAMESPACE=tyk
    APISecret=foo
    TykVersion=v5.3.0
    REDIS_BITNAMI_CHART_VERSION=19.0.2

    helm upgrade tyk-redis oci://registry-1.docker.io/bitnamicharts/redis -n $NAMESPACE --create-namespace --install --version $REDIS_BITNAMI_CHART_VERSION --set image.repository=bitnamilegacy/redis
    helm upgrade tyk-otel tyk-helm/tyk-oss -n $NAMESPACE --create-namespace \
    --install \
    --set global.secrets.APISecret="$APISecret" \
    --set tyk-gateway.gateway.image.tag=$TykVersion \
    --set global.redis.addrs="{tyk-redis-master.$NAMESPACE.svc.cluster.local:6379}" \
    --set global.redis.pass="$(kubectl get secret --namespace $NAMESPACE tyk-redis -o jsonpath='{.data.redis-password}' | base64 -d)" \
    --set tyk-gateway.gateway.opentelemetry.enabled=true \
    --set tyk-gateway.gateway.opentelemetry.exporter="grpc" \
    --set tyk-gateway.gateway.opentelemetry.endpoint="jaeger-all-in-one-collector.observability.svc:4317"
    ```

    <br />

    
<Note>

> Please make sure you are installing Redis versions that are supported by Tyk. Please refer to Tyk docs to get list of [supported versions](/tyk-self-managed/install#redis).

Reviewer: ```suggestion
    Please make sure you are installing Redis versions that are supported by Tyk. Please refer to Tyk docs to get list of [supported versions](/tyk-self-managed/install#redis).
```

</Note>

>     Tyk Gateway is now accessible through service gateway-svc-tyk-oss-tyk-gateway at port 8080 and exports the OpenTelemetry traces to the `jaeger-all-in-one-collector` service.

Reviewer: ```suggestion
    Tyk Gateway is now accessible through service `gateway-svc-tyk-oss-tyk-gateway` at port `8080` and exports the OpenTelemetry traces to the `jaeger-all-in-one-collector` service.
```

3. **Deploy Tyk Operator**

    Deploy Tyk Operator to manage APIs in your cluster:

    ```bash
    kubectl create namespace tyk-operator-system
    kubectl create secret -n tyk-operator-system generic tyk-operator-conf \
    --from-literal "TYK_AUTH=$APISecret" \
    --from-literal "TYK_ORG=org" \
    --from-literal "TYK_MODE=ce" \
    --from-literal "TYK_URL=http://gateway-svc-tyk-otel-tyk-gateway.tyk.svc:8080" \
    --from-literal "TYK_TLS_INSECURE_SKIP_VERIFY=true"
    helm install tyk-operator tyk-helm/tyk-operator -n tyk-operator-system

    ```

4. **Deploy a Test API Definition**

    Save the following API definition as `apidef-hello-world.yaml`:

    ```yaml
    apiVersion: tyk.tyk.io/v1alpha1
    kind: ApiDefinition
    metadata:
      name: hello-world
    spec:
      name: hello-world
      use_keyless: true
      protocol: http
      active: true
      proxy:
        target_url: http://httpbin.org/
        listen_path: /hello-world
        strip_listen_path: true
    ```

    To apply this API definition, run the following command:

    ```bash
    kubectl apply -f apidef-hello-world.yaml 
    ```

>     This step deploys an API definition named hello-world using the provided configuration. It enables a keyless HTTP API proxying requests to http://httpbin.org/ and accessible via the path /hello-world.

Reviewer: ```suggestion
    This step deploys an API definition named *hello-world* using the provided configuration. It enables a keyless HTTP API proxying requests to `http://httpbin.org/` and accessible via the path `/hello-world`.
```

5. **Explore OpenTelemetry traces in Jaeger**

    You can use the kubectl `port-forward command` to access Tyk and Jaeger services running in the cluster from your local machine's localhost:

>     For Tyk API Gateway:

Reviewer: ```suggestion
    For Tyk Gateway:
```

    ```bash
    kubectl port-forward service/gateway-svc-tyk-otel-tyk-gateway 8080:8080 -n tyk
    ```

    For Jaeger:

    ```bash
    kubectl port-forward service/jaeger-all-in-one-query 16686 -n observability
    ```

    Begin by sending a few requests to the API endpoint configured in step 2: 

    ```bash
    curl http://localhost:8080/hello-world/ -i
    ```

    Next, navigate to Jaeger on `http://localhost:16686`, select the ´service´ called ´tyk-gateway´ and click on the button ´Find traces´. You should see traces generated by Tyk:

>     <img src="/img/distributed-tracing/opentelemetry/api-gateway-trace-tyk-jaeger.png" alt="Tyk API Gateway distributed trace in Jaeger" />

Reviewer: ```suggestion
    <img src="/img/distributed-tracing/opentelemetry/api-gateway-trace-tyk-jaeger.png" alt="Tyk Gateway distributed trace in Jaeger" />
```

    Click on a trace to view all its internal spans:

>     <img src="/img/distributed-tracing/opentelemetry/api-gateway-trace-tyk-jaeger-spans.png" alt="Tyk API Gateway spans in Jaeger" />

Reviewer: ```suggestion
    <img src="/img/distributed-tracing/opentelemetry/api-gateway-trace-tyk-jaeger-spans.png" alt="Tyk Gateway spans in Jaeger" />
```
